src/pages/home/report/ReportActionsView.tsx

```diff
@@ -218,16 +219,62 @@ function ReportActionsView({
         [allReportActions, transactionThreadReportActions, transactionThreadReport?.parentReportActionID],
     );
 
-    const visibleReportActions = useMemo(
-        () =>
-            reportActions.filter(
-                (reportAction) =>
-                    (isOffline || isDeletedParentAction(reportAction) || reportAction.pendingAction !== CONST.RED_BRICK_ROAD_PENDING_ACTION.DELETE || reportAction.errors) &&
-                    shouldReportActionBeVisible(reportAction, reportAction.reportActionID, canPerformWriteAction, policy) &&
-                    isIOUActionMatchingTransactionList(reportAction, reportTransactionIDs),
-            ),
-        [reportActions, isOffline, canPerformWriteAction, reportTransactionIDs, policy],
-    );
+    const [allTransactions] = useOnyx(ONYXKEYS.COLLECTION.TRANSACTION, {canBeMissing: true});
+
+    const visibleReportActions = useMemo(() => {
+        const filtered = reportActions.filter((reportAction) => {
+            const passesBasicFilters =
+                (isOffline || isDeletedParentAction(reportAction) || reportAction.pendingAction !== CONST.RED_BRICK_ROAD_PENDING_ACTION.DELETE || reportAction.errors) &&
+                shouldReportActionBeVisible(reportAction, reportAction.reportActionID, canPerformWriteAction, policy);
+
+            if (!passesBasicFilters) {
+                return false;
+            }
+
+            if (isOffline && isMoneyRequestAction(reportAction)) {
+                return true;
+            }
+
+            const originalMessage = isMoneyRequestAction(reportAction) ? getOriginalMessage(reportAction) : undefined;
+            const iouReportID = originalMessage?.IOUReportID;
+            const transactionID = originalMessage?.IOUTransactionID;
+            let transactionIDsToCheck = reportTransactionIDs;
+
+            if (iouReportID && transactionID) {
+                const transactionInOnyx = allTransactions?.[`${ONYXKEYS.COLLECTION.TRANSACTION}${transactionID}`];
+
+                const iouReport = getReportOrDraftReport(iouReportID);
+                const isIOUReportExpense = isExpenseReport(iouReport);
+
+                if (isIOUReportExpense) {
+                    const expenseReportTransactions = Object.values(allTransactions ?? {}).filter((transaction) => {
```

Comment:
There is `allTransactions` on every filter iteration `reportActions.filter((reportAction)` and it looks quite complex. Maybe it would be better to create some utility function for it or even few of them (with tests ideally if possible)? I think it will be hard to test and maintain in this shape